flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/KMeans.java

```diff
@@ -153,39 +149,24 @@ public IterationBodyResult process(
             DataStream<Integer> terminationCriteria =
                     centroids.flatMap(new TerminateOnMaxIter(maxIterationNum));
 
-            DataStream<Tuple2<Integer, DenseVector>> centroidIdAndPoints =
+            DataStream<Tuple2<Integer[], DenseVector[]>> centroidIdAndPoints =
                     points.connect(centroids.broadcast())
                             .transform(
-                                    "SelectNearestCentroid",
+                                    "CentroidsUpdateAccumulator",
                                     new TupleTypeInfo<>(
-                                            BasicTypeInfo.INT_TYPE_INFO,
-                                            DenseVectorTypeInfo.INSTANCE),
-                                    new SelectNearestCentroidOperator(distanceMeasure));
+                                            BasicArrayTypeInfo.INT_ARRAY_TYPE_INFO,
+                                            ObjectArrayTypeInfo.getInfoFor(
+                                                    DenseVectorTypeInfo.INSTANCE)),
+                                    new CentroidsUpdateAccumulator(distanceMeasure));
```

Comment:
One optimization from algorithm side could be pre-compute the norm of each data point and thus accelerate the computing of `findClosestCentroidId`.

A similar solution could be found [1] [2].

[1] https://github.com/alibaba/Alink/blob/9ae07329a9fb77e39c334c09c79e5016b62f59ee/core/src/main/java/com/alibaba/alink/operator/common/distance/FastDistance.java
[2] https://github.com/apache/spark/blob/2fa98a6a0971cafeb98a2148a962cf8d79381842/mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala#L222